### ISSUE TITLE
Version 1.1.394 and 1.1.395 broke pandas-stubs

### ISSUE THREAD

**Issue by Dr-Irv** [microsoft/pyright#9995]
**Describe the bug**
The tests we run for `pandas-stubs` work fine with pyright 1.1.393 but fail with 1.1.394 and 1.1.395

**Code or Screenshots**

Set up `pandas-stubs` development environment as described here:
https://github.com/pandas-dev/pandas-stubs/blob/main/docs/setup.md

Use `poe pyright` to see the failures

**VS Code extension or command-line**

Command line and within VS Code

I thought you had tests that checked the `pandas-stubs` tests when you update `pyright`

**Comment by erictraut** [comment]
Could you please provide a simplified stand-alone code sample that demonstrates the issue?

**Comment by erictraut** [comment]
I try hard not to introduce regressions that affect major libraries. It's possible that I missed something here. It's also possible that your project is relying on behaviors and assumptions that have recently been clarified in the typing spec. I'm happy to work with you in finding the root cause of the problem, but I'll need you to provide more details.

**Comment by Dr-Irv** [comment]
> Could you please provide a simplified stand-alone code sample that demonstrates the issue?

It seems all the errors that are reported have to do with changes to how `@overload` is processed.  Might be related to #9984 

Isn't it the case that you run `mypy_primer` in your CI as part of the tests, and that it tests `pandas-stubs` ?  I thought you would pick up those changes as part of your testing methodology.

Here's a simple example:

File `overloadbug.py`:
```python
from ovb.overloadbugpyi import Foo


foo = Foo()
reveal_type(foo.abc())
```
File `ovb/__init__.py` is empty
File `ovb/overloadbug.pyi`:
```python
from typing import overload

class Foo:
    def __init__(self):
        pass

    @overload
    def abc(  # pyright: ignore[reportOverlappingOverload]
        self, index: bool = ..., name: str = ...
    ) -> tuple: ...
    @overload
    def abc(self, index: bool = ..., name: None = None) -> list: ...
```

Command line:   `pyright ovb overloadbug.py`

Version 1.1.393:
```text
overloadbug.py:25:13 - information: Type of "foo.abc()" is "tuple[Unknown, ...]"
```

Version 1.1.395:
```text
overloadbug.py:5:13 - information: Type of "foo.abc()" is "Unknown"
```

Result should be a `tuple`.  I guessed this because of #9984 .

But there are 180 errors in `pandas-stubs` with 1.1.395 and none with 1.1.393, so there is some regression.  A spot check seemed to show they are all on overloaded methods where we check the result with `assert_type()`

**Comment by Dr-Irv** [comment]
Based on the analysis here:  https://github.com/hauntsaninja/mypy_primer/issues/133#issuecomment-2689432556 it appears that this PR #9922 is what causes all the new features. 

The `mypy_primer` log here https://github.com/microsoft/pyright/pull/9922#issuecomment-2660630026  omitted all the `pandas-stubs` features that #9922 broke.

**Comment by erictraut** [comment]
OK, thanks for the additional analysis. The change in question was intentional and based on my reading of the typing spec, although it is arguably ambiguous in this particular case. I'll start a thread in the typing forum and see if we can get it to be spelled out more clearly in the typing spec.

This change doesn't affect type stub files, so it doesn't affect the pandas-stubs themselves. Depending on the decision that we reach on this issue, you may need to make changes in your pandas-stubs test code.

This issue is already tracked by https://github.com/microsoft/pyright/issues/9984, so I'm going to close this one as a duplicate.

**Comment by erictraut** [comment]
I've started [this discussion](https://discuss.python.org/t/default-argument-with-incompatible-type/82552) in the Python typing forum that relates to this issue. You're welcome to join that discussion if you have an opinion on the topic.

**Comment by MaffooBristol** [comment]
To avoid signing up on the python forums, I'd just like to say that I was so very confused about why my local pyright was giving 100 errors and my CI runners suddenly started giving 500 errors, turns out that loads and loads of libraries use that breaking default syntax. This is the kind of thing that I think should not be a patch release as it's so wildly different, or at least should be easily toggleable

**Comment by erictraut** [comment]
@MaffooBristol, thanks for the additional info. By any chance, can you provide other examples of libraries that are using this? If it is more widespread than just pandas, it's important for us to take that into consideration for this discussion.

**Comment by Dr-Irv** [comment]
As I wrote in the discussion, if I'm writing application code that is fully typed, and I don't want to create `.pyi` files, I'd like to write something like this:
```python
    @overload
    def get_variables(self, var_type: str) -> str: ...

    @overload
    def get_variables(self, var_type: Literal[None] = ...) -> dict[str, int]: ...

    def get_variables(
        self, var_type: str | None = None
    ) -> str | dict[str, int]:
        # Code here
```

I don't see how to do this if you don't interpret the ellipses in the overload declaration the same way that they are interpreted in stub files.

**Comment by erictraut** [comment]
> I don't see how to do this if you don't interpret the ellipses in the overload declaration the same way that they are interpreted in stub files. It's not just a library issue.

```python
from typing import overload

@overload
def get_variables(self, var_type: str) -> str: ...
@overload
def get_variables(self, var_type: None = None) -> dict[str, int]: ...

def get_variables(self, var_type: str | None = None) -> str | dict[str, int]: ...
```

**Comment by Dr-Irv** [comment]
OK, but putting default values in overloads isn't something I've seen a lot.  If you look at the docs for `mypy` at https://mypy.readthedocs.io/en/stable/more_types.html , there is this text:

The default values of a function’s arguments don’t affect its signature – only the absence or presence of a default value does. So in order to reduce redundancy, it’s possible to replace default values in overload definitions with ... as a placeholder:

```python
from typing import overload

class M: ...

@overload
def get_model(model_or_pk: M, flag: bool = ...) -> M: ...
@overload
def get_model(model_or_pk: int, flag: bool = ...) -> M | None: ...

def get_model(model_or_pk: int | M, flag: bool = True) -> M | None:
    ...
```

So I think lots of people may be using ellipses in overload declarations as a result of this suggestion in the `mypy` docs.

**Comment by erictraut** [comment]
I've published a hotfix release of pyright (1.1.396) to address this issue. I reverted the previous bug fix that caused this regression. The new version is already published to npm and should be available in the VS Code marketplace in the next few minutes and in pypi within the next 24 hours. Apologies for any inconvenience this caused.

**Comment by Dr-Irv** [comment]
> I've published a hotfix release of pyright (1.1.396) to address this issue. I reverted the previous bug fix that caused this regression. The new version is already published to npm and should be available in the VS Code marketplace in the next few minutes and in pypi within the next 24 hours. Apologies for any inconvenience this caused.

Thank you for doing this!  Also, I'd like to make a couple of suggestions:

1. If your CI that runs `mypy_primer` indicates an issue with `pandas-stubs`, and you're sure this is a problem that we need to fix, please create an issue in `pandas-stubs` so we can address it.  I think we will need to make a fix with respect to changes included in pyright 1.1.394, but I'll wait to address those until `1.1.396` is in `pypi`
2. As I indicated here:  https://github.com/microsoft/pyright/issues/9995#issuecomment-2690658704 - since it seems that the CI with `mypy_primer` could omit a report on `pandas-stubs` due to many issues being reported, it might be in our mutual best interest for such reports to be more visible, at least to say "180 new issues with `pandas-stubs`".  My sense is that if you saw 180 new issues with `pandas-stubs`, you might have investigated further before merging that change and making a new release.

**Comment by MaffooBristol** [comment]
Hi @erictraut - apologies, rushed my comment in as it was EOP Friday.

Discovered I was being slightly hyperbolic about it being "loads and loads" of packages. Turns out it's only two, one of which is [pdf2image](https://pypi.org/project/pdf2image/) and the other, ironically, [typed-ffmpeg](https://github.com/livingbio/typed-ffmpeg).

I've contributed to the latter and it's being actively maintained so could likely make the change there, and the former one seems quite dead so maybe deserves a typed fork anyway. I think the sheer number of errors generated for each line that the packages are used in meant it looked like it was wider spread than that, since it doesn't combine all the failing arguments into one.
